examples/src/chains/map_reduce_lcel.ts

```diff
@@ -0,0 +1,163 @@
+import { BaseCallbackConfig } from "langchain/callbacks";
+import {
+  _collapseDocs,
+  _splitListOfDocs,
+} from "langchain/chains/combine_documents/reduce";
+import { ChatOpenAI } from "langchain/chat_models/openai";
+import { Document } from "langchain/document";
+import { PromptTemplate } from "langchain/prompts";
+import { StringOutputParser } from "langchain/schema/output_parser";
+import { formatDocument } from "langchain/schema/prompt_template";
+import {
+  RunnableLambda,
+  RunnableMap,
+  RunnablePassthrough,
+  RunnableSequence,
+} from "langchain/schema/runnable";
+
+// Initialize the OpenAI model
+const model = new ChatOpenAI({});
+
+// Define prompt templates for document formatting, summarizing, collapsing, and combining
+const documentPrompt = PromptTemplate.fromTemplate("{pageContent}");
+const summarizePrompt = PromptTemplate.fromTemplate(
+  "Summarize this content:\n\n{context}"
+);
+const collapsePrompt = PromptTemplate.fromTemplate(
+  "Collapse this content:\n\n{context}"
+);
+const combinePrompt = PromptTemplate.fromTemplate(
+  "Combine these summaries:\n\n{context}"
+);
+
+// Define a util function to allow function invocation with preset arguments
+function partial<F extends (...args: any[]) => any>(
+  fn: F,
+  ...partialArgs: any[]
+): (...args: any[]) => ReturnType<F> {
+  return (...args: any[]): ReturnType<F> => fn(...partialArgs, ...args);
+}
+
+// Wrap the `formatDocument` util so it can format a list of documents
+const formatDocs = async (documents: Document[]): Promise<string> => {
+  const formattedDocs = await Promise.all(
+    documents.map((doc) => formatDocument(doc, documentPrompt))
+  );
+  return formattedDocs.join("\n\n");
+};
+
+// Define a function to get the number of tokens in a list of documents
+const getNumTokens = async (documents: Document[]): Promise<number> =>
+  model.getNumTokens(await formatDocs(documents));
+
+// Initialize the output parser
+const outputParser = new StringOutputParser();
+
+// Define the map chain to format, summarize, and parse the document
+const mapChain = RunnableSequence.from([
+  { context: async (i: Document) => formatDocument(i, documentPrompt) },
+  summarizePrompt,
+  model,
+  outputParser,
+]);
+
+// Define a wrapper chain to keep the original metadata while summarizing the document
+const mapAsDocChain = RunnableSequence.from([
```

Comment:
Can we just work this into `mapChain` to make it a bit easier to componentize?

Reply:
Yes, however I think what I had before was easier to comprehend.
New:
```typescript
// Define the final map-reduce chain
const mapReduceChain = RunnableSequence.from([
  RunnableSequence.from([
    RunnableMap.from({ doc: new RunnablePassthrough(), content: mapChain }),
    RunnableLambda.from(
      (input) =>
        new Document({ pageContent: input.content, metadata: input.doc.metadata })
    ),
  ]).withConfig({ runName: "Summarize (return doc)" }).map(),
  collapse,
  reduceChain,
]).withConfig({ runName: "Map reduce" });
```

Old:
```typescript
// Define a wrapper chain to keep the original metadata while summarizing the document
const mapAsDocChain = RunnableSequence.from([
  RunnableMap.from({ doc: new RunnablePassthrough(), content: mapChain }),
  RunnableLambda.from(
    (input) =>
      new Document({ pageContent: input.content, metadata: input.doc.metadata })
  ),
]).withConfig({ runName: "Summarize (return doc)" });

// Define the final map-reduce chain
const mapReduceChain = RunnableSequence.from([
  mapAsDocChain.map(),
  collapse,
  reduceChain,
]).withConfig({ runName: "Map reduce" });
```

Reply:
Ok yeah agreed

Reply:
What about:
```typescript
const mapChain =  RunnableSequence.from([
    { doc: new RunnablePassthrough(), content: mapChain },
     (input) => new Document({ pageContent: input.content, metadata: input.doc.metadata }),
  ]).withConfig({ runName: "Summarize (return doc)" }).map(),
```

Reply:
Yeah removing `RunnableLambda` works, pushed update